### PR TITLE
Return Vec<U> from queries instead of <Box<Iterator<Item=T>>

### PR DESCRIPTION
**Pull request by mfpiccolo**
This at least partially addresses #132.  @sgrif @mcasper

### CODE REVIEW

**Comment by sgrif** [comment]
This looks fine. I'm going to hold off on merging until we get a bit closer to release, as I'd just like to think about it a bit more to be 100% sure this is the right thing to do here.

**Comment by sgrif** [comment]
This will need a changelog entry (go ahead and change the entry about `Iterator` since that's still unreleased)

**Comment by sgrif** [comment]
This PR also needs to ensure that errors returned by `FromSql`, or `FromSqlRow` do not result in a `panic!`. We should have a test to verify this (probably something abusing the `NativeSqlType` node, as actually causing one of these errors should basically be impossible using the query DSL)

File: diesel/src/connection/mod.rs [hunk 1]
```diff
@@ -78,11 +78,11 @@ pub trait Connection: SimpleConnection + Sized {
         U: Queryable<T::SqlType, Self::Backend>,
     {
         self.query_all(source)
-            .and_then(|mut e| e.nth(0).map(Ok).unwrap_or(Err(Error::NotFound)))
+            .and_then(|e: Vec<U>| e.into_iter().next().ok_or(Error::NotFound))
```
Comment (sgrif) [inline diesel/src/connection/mod.rs]: Should this be `.and_then(|e| e.first().map(Ok).unwrap_or(Err(Error::NotFound)))`?

File: diesel/src/connection/mod.rs [hunk 2]
```diff
@@ -78,11 +78,11 @@ pub trait Connection: SimpleConnection + Sized {
...
     }
 
     #[doc(hidden)]
-    fn query_all<'a, T, U: 'a>(&self, source: T) -> QueryResult<Box<Iterator<Item=U> + 'a>> where
+    fn query_all<'a, T, U>(&self, source: T) -> QueryResult<Vec<U>> where
```
Comment (sgrif) [inline diesel/src/connection/mod.rs]: You can remove this lifetime.

File: diesel/src/connection/pg/mod.rs [hunk 3]
```diff
@@ -57,7 +57,7 @@ impl Connection for PgConnection {
         self.execute_inner(query).map(|res| res.rows_affected())
     }
 
-    fn query_all<'a, T, U: 'a>(&self, source: T) -> QueryResult<Box<Iterator<Item=U> + 'a>> where
+    fn query_all<'a, T, U>(&self, source: T) -> QueryResult<Vec<U>> where
```
Comment (sgrif) [inline diesel/src/connection/pg/mod.rs]: You can remove this lifetime.

File: diesel/src/migrations/mod.rs [hunk 4]
```diff
@@ -145,10 +144,9 @@ pub fn create_schema_migrations_table_if_needed<Conn: Connection>(conn: &Conn) -
     })
 }
 
-fn previously_run_migration_versions<Conn: Connection>(conn: &Conn) -> QueryResult<HashSet<String>> {
+fn previously_run_migration_versions<Conn: Connection>(conn: &Conn) -> QueryResult<Vec<String>> {
     __diesel_schema_migrations.select(version)
         .load(conn)
-        .map(|r| r.collect())
```
Comment (sgrif) [inline diesel/src/migrations/mod.rs]: Ugh. See this sucks. I really liked being able to stick it into a `HashSet` like this. :\

Comment (sgrif) [inline diesel/src/migrations/mod.rs]: This should still return a `HashSet` regardless. This line should be `.map(HashSet::from_iter)`

File: diesel/src/connection/mod.rs (hunk 1 above)
Comment (mfpiccolo) [inline diesel/src/connection/mod.rs]: `first()` returns `&U` so that would give us `Result<&U, Error>` instead of `Result<U, Error>`

**Comment by sgrif** [comment]
I rebased and finished this PR manually.

**Comment by mfpiccolo** [comment]
:heart: I was messing with this last night trying to figure out a good solution for HashSet but didn't get there.  I didn't realize that that method was already changed.

**Comment by sgrif** [comment]
I fixed it as part of the rebase.

**Pull request merged**